apps/astarte_pairing_api/lib/astarte_pairing_api_web/router.ex

```diff
@@ -22,6 +22,13 @@ defmodule Astarte.Pairing.APIWeb.Router do
   pipeline :api do
     plug :accepts, ["json"]
     plug Astarte.Pairing.APIWeb.Plug.LogRealm
+    plug Astarte.Pairing.APIWeb.Plug.AuthorizePath
+  end
+
+  pipeline :api_device do
+    plug :accepts, ["json"]
+    plug Astarte.Pairing.APIWeb.Plug.LogRealm
+    plug Astarte.Pairing.APIWeb.Plug.LogHwId
```

Comment:
I would remove `plug :accepts, ["json"]` and `plug Astarte.Pairing.APIWeb.Plug.LogRealm`, since they are already applied by the `:realm_api` pipeline.

Let's use scopes to appropriately structure the routes and avoid applying the plugs multiple times.
The structure suggested by Arnaldo seems fine, was there some issues with it?

```elixir
  pipeline :realm_api do
    plug :accepts, ["json"]
    plug Astarte.Pairing.APIWeb.Plug.LogRealm
  end

  pipeline :agent_api do
    plug Astarte.Pairing.APIWeb.Plug.AuthorizePath
  end

  pipeline :devices_api do
    plug Astarte.Pairing.APIWeb.Plug.LogHwId
  end

  scope "/v1/:realm_name", Astarte.Pairing.APIWeb do
    pipe_through :realm_api

    get "/version", VersionController, :show

    scope "/agent" do
      pipe_through :agent_api

      post "/devices", AgentController, :create
      delete "/devices/:device_id", AgentController, :delete
    end

    scope "/devices" do
      pipe_through :devices_api

      get "/:hw_id", DeviceController, :show_info
      post "/:hw_id/protocols/:protocol/credentials", DeviceController, :create_credentials
      post "/:hw_id/protocols/:protocol/credentials/verify",
           DeviceController,
           :verify_credentials
    end
  end
```